### PR TITLE
Makes auxiliary task independent of dynamic models

### PR DESCRIPTION
**Pull request by lucasosouza**
I made a recent change to try to reduce how many times `update_auxiliary_task` is called, based on the comments you left on the code. 

Now there are 4 places in the code where `update_auxiliary_task` is being called :
- after collecting the rollout and before the reward is calculated
- in the ppo loss, it is called once, before the dynamic loss
- in the backprop loss, it is called twice - before the dynamic loss and before the backprop loss. The reason being that I'm considering there is a gradient step after the dynamic loss, which would change the model and affect the output of `update_auxiliary_task`

If the last statement is incorrect, and `update_auxiliary_task` needs to be called only once per step/epoch/batch, then we can move out of the loss functions back into the `learn` function and both the ppo loss and backprop loss would share that procedure. 

At some point we can look into the `update_auxiliary_task` as well to see if it can be further simplified. 

Edit: I changed the name of the function `ppo.update` to `ppo.learn` to make it clear that is where the gradient steps are happening. There are lots of update functions around the code, and I wanted to differentiate that is the only one that actually has a gradient update. Let me know if that is not ok.

### CODE REVIEW

File: nupic/embodied/agents/curious_ppo_agent.py [hunk 1]
```diff
@@ -508,76 +525,90 @@ def log_post_update(self, info):
...
             "loss/auxiliary_task": aux_loss
         }
 
-    def backprop_loss(self, acs, obs, last_obs):
-        pred_features = []
-        # Get output from all dynamics models (featurewise)
-        # shape=[num_dynamics, num_envs, n_steps_per_seg, feature_dim]
+    def update_auxiliary_task(self, acs, obs, last_obs, return_next_features=True):
+        # Update the auxiliary task
+        self.auxiliary_task.policy.update_features(obs, acs)
```
Comment (vkakerbeck) [inline nupic/embodied/agents/curious_ppo_agent.py:543]: I think we could skip this here if we are not sharing features with the policy. This would save us a forward pass through the policy network. If we don't call policy.update_features we have to remember to pass the current action batch to the auxiliary_task update or simply remove acs altogether from this function since there is no change happening on it (asc==ac)

**Review by vkakerbeck — APPROVED** [review]
It looks very good. I think in a few places we can still optimize and take out a forward pass from the update_auxiliary_task function (I added a comment in the line).

As far as I see you got all the places where we need to call update_auxiliary_task. I think we could avoid calling it twice in the backprob loss. There would be some small change in the features after the first update using the dynamic loss so to capture that we would have to call it twice but I don't know if the effect of that change outweighs the advantages of only calling it once and then using the old features for the dissagreement update as well. I don't have a strong opinion on that.

The name change is a good idea. We could also change functions that are called update_x to get_x and/or set_x since no updates happen there. We only get features by performing a forward pass and set the class variables to the new features. If you think that would be more clear we could also change those names.
